### ISSUE TITLE
convert_sync_batchnorm should respect device affinity

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#37930]
`torch.nn.SyncBatchNorm.convert_sync_batchnorm` by default places newly created parameters on CPU even if all parameters of input model are on GPU. It might be better to respect input module's device affinity and place new parameters on the same device where the original `_BatchNorm` module resides.

To reproduce:

```python
import torch
module = torch.nn.Sequential(
        torch.nn.Linear(20, 100),
        torch.nn.BatchNorm1d(100)
).cuda()
print(set([p.device for p in module.parameters()]))
sync_bn_module = torch.nn.SyncBatchNorm.convert_sync_batchnorm(module)
print(set([p.device for p in sync_bn_module.parameters()]))
```

The output is:

```
{device(type='cuda', index=0)}
{device(type='cuda', index=0), device(type='cpu')}
```

cc @albanD @mruberry

**Comment by mrshenli** [comment]
Adding triage review add to discuss whether the current behavior is expected or should be fixed.

**Comment by rohan-varma** [comment]
To add some context, we're seeing this issue come up for DDP users where the typical flow is something like:
```
model.to(cuda_idx)
model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)
model = DDP(model, device_ids=[cuda_idx])
```

The error that's raised by DDP here indicates that sync BN ended up placing some parameters back on the CPU, even though the model was moved to cuda above. The current workaround is to move to cuda after converting to sync BN, but I don't think the order should have to matter.

**Comment by soumith** [comment]
i looked at this, i'm pretty sure this `sync_bn_module` has to be on `cuda`, because it's converting `module` which is on `cuda`

**Comment by jjsjann123** [comment]
Looks like the assignment here ended up not making a copy of tensor any more: https://github.com/pytorch/pytorch/commit/79533ef097bf3ff2f1d85782ae1f279fe15e8686#diff-98c9603e09a23834975e9d62c07c30d9R498-R499

This is very strange, there is a change of behavior in pytorch...did we rewrite the set attribute thing at some point?

**Comment by jjsjann123** [comment]
oops, I was looking at the wrong thing...
It's the script update, we are calling `copy_` in the scripts that's breaking the device. :https://github.com/pytorch/pytorch/blob/cd0724f9f1b57dae12be2c3fc6be1bd41210ee88/torch/nn/modules/batchnorm.py#L501-L502

**Comment by orchidmajumder** [comment]
This seems to be a breaking change between `1.4` and `1.5` because I started facing this issue just after the upgrade. Will this be reverted back in the next release?

**Comment by rwightman** [comment]
I also just ran into this on 1.5, wasn't an issue in 1.4.

**Comment by jjsjann123** [comment]
Fixed in PR #38729
